provider/inmemkv/main.go

```diff
@@ -0,0 +1,122 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	provider "github.com/wasmCloud/provider-sdk-go"
+	core "github.com/wasmcloud/interfaces/core/tinygo"
+)
+
+type data map[string]interface{}
+
+var (
+	p        *provider.WasmcloudProvider
+	database map[string]data
+	lock     sync.RWMutex
+
+	ErrNotImplemented   error = errors.New("operation not implemented")
+	ErrInvalidOperation error = errors.New("operation not valid")
+)
+
+func main() {
+	var err error
+	database = make(map[string]data)
+	lock = sync.RWMutex{}
+
+	p, err = provider.New(
+		"wasmcloud:keyvalue",
+		provider.WithProviderActionFunc(handleKVAction),
+		provider.WithNewLinkFunc(handleNewLink),
+		provider.WithDelLinkFunc(handleDelLink),
+		provider.WithHealthCheckMsg(healthCheckMsg),
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	err = p.Start()
+	if err != nil {
+		panic(err)
+	}
+}
+
+func healthCheckMsg() string {
+	return fmt.Sprintf("Num databases initialized: %d", len(database))
+}
+
+func handleDelLink(linkdef core.LinkDefinition) error {
+	delete(database, linkdef.ActorId)
+	return nil
+}
+
+func handleNewLink(linkdef core.LinkDefinition) error {
+	newTable := make(map[string]interface{})
+	database[linkdef.ActorId] = newTable
+	return nil
+}
+
+func handleKVAction(a provider.ProviderAction) (*provider.ProviderResponse, error) {
+	resp := new(provider.ProviderResponse)
+	var err error
+
+	switch a.Operation {
+	case "KeyValue.Increment":
+		msg := decodeIncrementRequest(a.Msg)
+
+		// Get the actor specific database
+		db := database[a.FromActor]
+
+		if db[msg.Key] == nil {
+			lock.Lock()
+			db[msg.Key] = msg.Value
+			lock.Unlock()
+		} else {
+			value, ok := db[msg.Key].(int32)
+			if !ok {
+				return nil, errors.New("key value not of int type")
+			}
+
+			lock.Lock()
+			db[msg.Key] = value + msg.Value
+			lock.Unlock()
+		}
+
+		resp.Msg, err = encodeIncrementResponse(db[msg.Key].(int32))
+		if err != nil {
+			return nil, err
+		}
+		return resp, nil
+	case "KeyValue.Contains":
+		return nil, ErrNotImplemented
+	case "KeyValue.Del":
```

Comment:
@jordan-rash any chance we could get set and delete as well as just increment?

Reply:
yes sir....ill get that done by end of week.

Reply:
End of which week 👀 

@jordan-rash came across this one again since I wanted to use it as an example, so just bumping since this looks good. If we could get set/get/del/increment that's fantastic.

No big rush

Reply:
I feel like "week" is a more abstract construct that we, as a human race, have pigeon holed into an obscure 168 hour window.